openedx/license-manager
# feat: add license_duration_before_purge field to CustomerAgreement

Add a `license_duration_before_purge` field to the CustomerAgreement model, so that we can configure the number of days before unused licenses are purged on a per-customer basis.  This change applies to all 3 cases that the `retire_old_licenses` command deals with:
1. licenses in plans whose expiration date was >=90 days ago and that have a non-null user_email
2. revoked licenses in any plan whose revocation date was >=90 days ago and that have a non-null user_email
3. assigned (but not activated) licenses whose assigned date was >=(90 by default, but now configurable per-customer) days ago.

https://openedx.atlassian.net/browse/ENT-4795

## Prologue

Note: doing timedelta-based math based on the value of a field in a single Django query that's backed by MySQL is tricky - you need to use an [`F()` expression](https://docs.djangoproject.com/en/2.2/topics/db/queries/#using-f-expressions-in-filters), but you can't provide an `F()` expression as an argument to a `timedelta()` object .  To make matters worse, you can't just use the trick of multiplying a `timedelta(days=1)` by the field value, because MySQL doesn't support `*` as a datetime arithmetic operator, i.e.
```python
assigned_date__date__lt=(
    localized_utcnow().date() - F('subscription_plan__customer_agreement__license_duration_before_purge') * timedelta(1)
)
```
doesn't work (it actually _would_ work if we were using Postgres).

So, I used a [`DurationField`](https://docs.djangoproject.com/en/2.2/ref/models/fields/#django.db.models.DurationField).
The problem with this type of field is that, in MySQL, the underlying DB value is stored as microseconds, which is a really bad experience in the Django Admin, because the value you'd have to enter for 90 days is actually (90 * 24 * 60 * 60 * 1000000).

That's why there's a new [`django-durationwidget`](https://github.com/devangpadhiyar/django-durationwidget) installed here - it makes the admin experience easier.

## Testing considerations

* Go to http://localhost:18170/admin/subscriptions/customeragreement/ and select a customer agreement
* Notice the license_duration_before_purge field.  Change it from the default
* Save and continue editing
* Notice the value changed.
![image](https://user-images.githubusercontent.com/2307986/128078449-6b607280-2f96-4009-96e3-23e9979ce529.png)

* Now go make some licenses that are assigned, with an assigned date prior to the number of days ago specified by the duration.
![image](https://user-images.githubusercontent.com/2307986/128078753-0a719cb4-8b94-4d92-b2dd-74141ab60efc.png)

* In an `app-shell`, run `./manage.py retire_old_licenses` - it should log something about retiring this license (it'll mention the UUID).
* Notice your assigned licenses in this agreement have now been scrubbed of user data and reset to unassigned.
## Post-review
